iothub_client/samples/iotedge_downstream_device_sample/iotedge_downstream_device_sample.c

```diff
@@ -0,0 +1,313 @@
+            if (result == NULL)
+            {
+                printf("Could not allocate memory to hold the certificate\r\n");
+            }
+            else
+            {
+                // copy the file into the buffer
+                size_t read_size = fread(result, 1, file_size - 1, ca_file);
+                if (read_size != file_size - 1)
+                {
+                    printf("Error reading file %s\r\n", edge_ca_cert_path);
+                    free(result);
+                    result = NULL;
+                }
+            }
+        }
+        (void)fclose(ca_file);
+    }
+
+    return result;
+}
+
+int main(void)
+{
+    IOTHUB_CLIENT_TRANSPORT_PROVIDER protocol;
+    IOTHUB_MESSAGE_HANDLE message_handle;
+    const char* telemetry_msg = "test_message";
+    char *cert_string = NULL;
+
+    printf("This sample will send %d messages and wait for any C2D messages.\r\nPress the enter key to end the sample\r\n\r\n", MESSAGE_COUNT);
+
+    // Select the Protocol to use with the connection
+#ifdef SAMPLE_MQTT
+    protocol = MQTT_Protocol;
+#endif // SAMPLE_MQTT
+#ifdef SAMPLE_MQTT_OVER_WEBSOCKETS
+    protocol = MQTT_WebSocket_Protocol;
+#endif // SAMPLE_MQTT_OVER_WEBSOCKETS
+#ifdef SAMPLE_AMQP
+    protocol = AMQP_Protocol;
+#endif // SAMPLE_AMQP
+#ifdef SAMPLE_AMQP_OVER_WEBSOCKETS
+    protocol = AMQP_Protocol_over_WebSocketsTls;
+#endif // SAMPLE_AMQP_OVER_WEBSOCKETS
+#ifdef SAMPLE_HTTP
+    protocol = HTTP_Protocol;
+#endif // SAMPLE_HTTP
+
+    IOTHUB_DEVICE_CLIENT_HANDLE device_handle;
+
+    // Used to initialize IoTHub SDK subsystem
+    (void)IoTHub_Init();
+
+    (void)printf("Creating IoTHub handle\r\n");
+    // Create the iothub handle here
+    device_handle = IoTHubDeviceClient_CreateFromConnectionString(connectionString, protocol);
+    if (device_handle == NULL)
+    {
+        (void)printf("Failure createing Iothub device.  Hint: Check you connection string.\r\n");
+    }
+    else
+    {
+        // Setting message callback to get C2D messages
+        (void)IoTHubDeviceClient_SetMessageCallback(device_handle, receive_msg_callback, NULL);
+        // Setting connection status callback to get indication of connection to iothub
+        (void)IoTHubDeviceClient_SetConnectionStatusCallback(device_handle, connection_status_callback, NULL);
+
+        // Set any option that are neccessary.
+        // For available options please see the iothub_sdk_options.md documentation
+
+        // Turn tracing on/off
+        // bool traceOn = true;
+        // (void)IoTHubDeviceClient_SetOption(iothub_handle, OPTION_LOG_TRACE, &traceOn);
+
+        // Provide the Azure IoT device client with the Edge root
+        // X509 CA certificate that was used to setup the Edge runtime
+        cert_string = obtain_edge_ca_certificate();
+        (void)IoTHubDeviceClient_SetOption(device_handle, OPTION_TRUSTED_CERT, cert_string);
+
+#if defined SAMPLE_MQTT || defined SAMPLE_MQTT_OVER_WEBSOCKETS
```

Comment:
this section has a bunch of commented out code inside #if block...   any reason it's commented out?

Reply:
@jspaith can correct me but the comments here make sense to me and are probably not too complicated for a general audience as well.